### ISSUE TITLE
Code performance overhaul 

### ISSUE THREAD

**Issue by nawtrey** [Becksteinlab/kda#2]
It was found that for large diagrams with many edges, the number of partial diagrams is very large (1e6 or more). With this number of diagrams the memory footprint of this code is large as well, and can easily consume Gb's or tens of Gb's of RAM. While the number of partial diagrams seems to scale as N!, there are some ways that might help reduce the memory footprint:
 - [x] Change functions to work in terms of lists of edge tuples instead of diagrams
 - [ ] ~~Save node labels as integers~~
 - [x] Find ways to reduce number of total graphs loaded in memory at a time
 - [ ] ~~Store values as float32 instead of float64~~

~~Note: no partial diagrams, directional partial diagrams, or flux diagrams contain edge data. They only contain node "istarget" data (True or False).~~

**Comment by nawtrey** [comment]
It should be noted that since the number of partial diagrams scale with N!, these measures will not address the real issue, but only soften the blow. It might be possible to reduce the memory footprint an order of magnitude or more, but if a system is sufficiently large this will basically do nothing.

**Comment by orbeckst** [comment]
Good to have an issue for this. 

I wouldn’t overly worry about it until it really becomes a problem and prevents you from obtaining results.

**Comment by nawtrey** [comment]
I've crossed out some of the recommendations because they are not substantial contributors to the memory footprint at this time. I've also checked off "reduce number of graphs loaded into memory" because it is kind of a non-issue since the standard is now to use arrays of edges instead of `NetworkX` diagrams. The primary contributor to the memory footprint was found to be the loading of all combinations of possible edge tuples (for generating all spanning trees), which was changed to use a generator. 

So with #25 being merged the memory footprint has been substantially reduced such that I think this issue can be closed. If memory becomes an issue again, new analysis will have to be done and new issues can be opened.
